app/controllers/supplemental_files_controller.rb

```diff
@@ -0,0 +1,122 @@
+    matching_file = @master_file.supplemental_files.find { |file| file.id == params[:id] }
+    # FIXME: redirect or proxy the content instead of rails directly sending the file
+    send_file matching_file.absolute_path
+  end
+
+  # Update the label of the supplemental file
+  def update
+    authorize! :edit, @master_file
+    file_index = @master_file.supplemental_files.find_index { |file| file.id == @supplemental_file.id }
+    raise Avalon::NotFound, "Cannot update the supplemental file: #{ @supplemental_file.id } not found" unless file_index.present?
+    raise Avalon::BadRequest, "Updating file contents not allowed" if supplemental_file_params[:file].present?
+
+    # Use the stored absolute path and not update from the user params
+    @supplemental_file.absolute_path = @master_file.supplemental_files[file_index].absolute_path
+    supplemental_files = @master_file.supplemental_files
+    supplemental_files[file_index] = @supplemental_file
+    @master_file.supplemental_files = supplemental_files
+    raise Avalon::SaveError, @master_file.errors[:supplemental_files_json].full_messages unless @master_file.save
+
+    respond_to do |format|
+      format.html do
+        flash[:success] = "Supplemental file successfully updated."
+        redirect_to edit_media_object_path(@master_file.media_object_id, step: 'structure')
+      end
+      format.json do
+        head :ok, location: master_file_supplemental_file_path(id: @supplemental_file.id, master_file_id: @master_file.id)
+      end
+    end
+  end
+
+  def destroy
+    authorize! :edit, @master_file
+    file = @master_file.supplemental_files.find { |file| file.id == params[:id] }
+    raise Avalon::NotFound, "Cannot update the supplemental file: #{ params[:id] } not found" unless file.present?
+    @master_file.supplemental_files -= [file]
+
+    raise Avalon::SaveError, "An error occurred when deleting the supplemental file: #{@master_file.errors[:supplemental_files_json].full_messages}" unless @master_file.save
+
+    respond_to do |format|
+      format.html do
+        flash[:success] = "Supplemental file successfully deleted."
+        redirect_to edit_media_object_path(@master_file.media_object_id, step: 'structure')
+      end
+      format.json do
+        head :no_content
+      end
+    end
+  end
+
+  private
+
+    def set_master_file
+      @master_file = MasterFile.find(params[:master_file_id])
+    end
+
+    def build_supplemental_file
+      # Note that built supplemental file does not have an absolute_path
+      @supplemental_file = MasterFile::SupplementalFile.new(id: params[:id], label: supplemental_file_params[:label])
+    end
+
+    def supplemental_file_params
+      # TODO: Add parameters for minio and s3
+      params.fetch(:supplemental_file, {}).permit(:label, :file)
+    end
+
+    def next_id
+      last_id = @master_file.supplemental_files.max_by { |f| f.id.to_i }
+      last_id.present? ? last_id + 1 : 1
```

Comment:
@cjcolvar I get an error here `undefined method '+' for #<MasterFile::SupplementalFile:0x00007effbaf2a128>` when adding a new file. Here `label_id` is the Supplemental_file object not the id as expected.
So I changed that line to the following to make it work;
```
last_id.present? ? last_id.id.to_i + 1 : 1
```

Reply:
Good catch.  I was trying to simplify that line and missed that change in return value.  I'll probably just revert that change, but it looks like I need more tests if this is failing and the tests didn't catch it.